### ISSUE TITLE
Add support for the "concealed" graphic rendition attribute

### ISSUE THREAD

**Issue by j4james** [microsoft/terminal#6876]
# Description of the new feature/enhancement

The ANSI `SGR 8` rendition attribute is used to indicate that text to which it's applied should be rendered invisible. I'm not sure how widely it's used, but it is supported by most terminal emulators, so I think it's probably worth doing. And most of the framework is already in place, so it's an easy addition for us.

# Proposed technical implementation details (optional)

I think all that's required is an update to the `TextAttribute::CalculateRgbColors` method, adjusting the colors to make the foreground and background the same when the attribute is set. Something like:

    if (IsInvisible())
    {
        fg = bg;
    }

In testing this on other terminals, I found that XTerm also prevented the invisible content being copied to the clipboard. However, pretty much everyone else allowed it to be copied, so I'm more inclined to follow the majority behaviour.

I have a PR ready to submit for this if you're happy with the idea and the proposed approach, but it'd be best if #6873 is merged first, otherwise there'll probably be conflicts in the unit tests.

**Comment by DHowett** [comment]
Philosophically, if it's not copyable and not rendered in xterm, is there any meaningful difference between it being concealed and it simply not existing?

For us, matching the colors and leaving it copyable sounds great. I'm r+ on this proposal!

**Comment by j4james** [comment]
> Philosophically, if it's not copyable and not rendered in xterm, is there any meaningful difference between it being concealed and it simply not existing?

That depends on the implementation. If they're still writing the text to the buffer, then it could potentially be revealed through other means, e.g. by changing the attributes with `DECARA`, or reading the content back via `DECRQCRA`. I couldn't get `DECARA` to work, but I'm not sure if that's just a bug. And I didn't try `DECRQCRA`, but I suspect it wouldn't work either - my guess is they're not actually storing anything.

Anyway I don't think it matters much. This is one of those things that doesn't have a clear right answer, because the ANSI/ECMA specs don't go into any detail, and the DEC terminals never had this functionality. I figure as long as we're doing what most others are doing then we're probably OK.

**Comment by jdebp** [comment]
`DECCARA` is documented as only affecting specific attributes, of which this is not one.  But the DEC doco does state that invisibility is rendered by making the foreground colour the same as the background colour.

Observe, furthermore, that kitty extends `DECCARA` to cover this attribute.

**Comment by j4james** [comment]
> But the DEC doco does state that invisibility is rendered by making the foreground colour the same as the background colour.

That's good to know - thanks for pointing that out. I was under the impression that none of the DEC terminals supported the concealed/invisible attribute, possibly because it's not listed in the functions summary table of the VT520 Programmer Reference. But I see now that even the VT420 supported it.

> `DECCARA` is documented as only affecting specific attributes, of which this is not one. 

Yeah I saw that, but I thought (incorrectly) that that was just because they didn't support the concealed attribute in general. I suppose it's still possible that was an oversight in the docs.

> Observe, furthermore, that kitty extends `DECCARA` to cover this attribute.

I thought XTerm did that too (see [here](https://github.com/ThomasDickey/xterm-snapshots/blob/153d0ccc03242cd2844255a7e03e8683a17132f3/screen.c#L2722-L2724)), but as I mentioned above I couldn't get it to work. I don't think I tested `DECCARA` on kitty. Not that it matters much at the moment. We can figure out those details when we actually implement those ops.
